### PR TITLE
Added doc for more string-dict methods

### PR DESCRIPTION
**Pull request by ds26gte**
This addresses brownplt/pyret-lang#1245. It adds documentation for the immutable string-dict methods `{merge, keys-list, map-keys, fold-keys, each-key}` and the mutable-string-dict methogs `{merge-now, keys-list-now, map-keys-now, fold-keys-now, each-key-now}`.

### CODE REVIEW

File: src/trove/string-dict.scrbl [hunk 1]
```diff
@@ -225,6 +236,102 @@ check:
...
 }
 
+@sd-method["merge"
+  #:contract (a-arrow (SD-of "a") (SD-of "a") (SD-of "a"))
+  #:args (list (list "self" #f) (list "other" #f))
+  #:return (SD-of "a")
+]
+
+Returns a new immutable string-dict that has the keys of both
+the original string-dict and the @pyret{other} string-dict. Choose a key's
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: I wouldn't use the verb "choose" here, since that sounds to me like the programmer has a choice.  I'd reword this to "In cases where both input dictionaries have the same key, the resulting value comes from the `@pyret{other}` string-dict."

File: src/trove/string-dict.scrbl [hunk 2]
```diff
@@ -225,6 +236,102 @@ check:
...
+
+@examples{
+check:
+  sd1 = [string-dict: "a", 5, "c", 4]
+  sd2 = [string-dict: "a", 10, "b", 6]
+  sd3 = sd1.merge(sd2)
+  sd3 is [string-dict: "a", 10, "b", 6, "c", 4]
+  sd4 = sd2.merge(sd1)
+  sd4 is [string-dict: "a", 5, "b", 6, "c", 4]
+  sd2.merge(sd1).merge(sd1) is sd2.merge(sd1)
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: `# merging the same dictionary twice has no additional effect`

File: src/trove/string-dict.scrbl [hunk 3]
```diff
@@ -225,6 +236,102 @@ check:
...
+otherwise it has the value from the original string-dict.
+
+@examples{
+check:
+  sd1 = [string-dict: "a", 5, "c", 4]
+  sd2 = [string-dict: "a", 10, "b", 6]
+  sd3 = sd1.merge(sd2)
+  sd3 is [string-dict: "a", 10, "b", 6, "c", 4]
+  sd4 = sd2.merge(sd1)
+  sd4 is [string-dict: "a", 5, "b", 6, "c", 4]
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: `# NOTE: sd3 is-not sd4`

File: src/trove/string-dict.scrbl [hunk 4]
```diff
@@ -225,6 +236,102 @@ check:
...
+end
+}
+
+@sd-method["keys-list"
+  #:contract (a-arrow (SD-of "a") (L-of S))
+  #:args (list (list "self" #f))
+  #:return (L-of S)
+]
+
+Returns the list of keys in the immutable string-dict, in some order.
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: "in some unspecified order"

File: src/trove/string-dict.scrbl [hunk 5]
```diff
@@ -225,6 +236,102 @@ check:
...
+
+@examples{
+check:
+  sd2 = [string-dict: "a", 10, "b", 6]
+  sd2.keys-list() is [list: "a", "b"]
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: If the order is unspecified, then this test might fail.  Either add a `.sort()` or make it into a set or something...

File: src/trove/string-dict.scrbl [hunk 6]
```diff
@@ -225,6 +236,102 @@ check:
...
+  #:contract (a-arrow (SD-of "a") (a-arrow "a" "b") (L-of "b"))
+  #:args (list (list "self" #f) (list "f" #f))
+  #:return (L-of "b")
+]
+
+Applies the function @pyret{f} to each key in the immutable string-dict, and
+returns a list of the returned values in some order.
+
+@examples{
+fun one-of(ans, elts):
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: Probably easier to say `elts.member(ans)`, no?

File: src/trove/string-dict.scrbl [hunk 7]
```diff
@@ -225,6 +236,102 @@ check:
...
+end
+}
+
+@sd-method["each-key"
+  #:contract (a-arrow (SD-of "a") (a-arrow S No) No)
+  #:args (list (list "self" #f) (list "f" #f))
+  #:return No
+]
+
+Calls the function @pyret{f} on each key in the immutable string-dict,
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: The point of this is the side effects of `f`, right?  Probably worth mentioning.

File: src/trove/string-dict.scrbl [hunk 8]
```diff
@@ -423,4 +530,102 @@ check:
...
 }
 
+@msd-method["merge-now"
+  #:contract (a-arrow (MSD-of "a") (MSD-of "a") No)
+  #:args (list (list "self" #f) (list "other" #f))
+  #:return No
+]
+
+Modifies the mutable string-dict to include the keys from the @pyret{other}
+mutable-string-dict. Choose a key's value to be the one it has in the @pyret{other}
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: ditto

File: src/trove/string-dict.scrbl [hunk 9]
```diff
@@ -423,4 +530,102 @@ check:
...
+end
+}
+
+@msd-method["keys-list-now"
+  #:contract (a-arrow (MSD-of "a") (L-of S))
+  #:args (list (list "self" #f))
+  #:return (L-of S)
+]
+
+Returns the list of keys in the mutable string-dict, in some order.
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: ditto

File: src/trove/string-dict.scrbl [hunk 10]
```diff
@@ -423,4 +530,102 @@ check:
...
+end
+}
+
+@msd-method["map-keys-now"
+  #:contract (a-arrow (MSD-of "a") (a-arrow "a" "b") (L-of "b"))
+  #:args (list (list "self" #f) (list "f" #f))
+  #:return (L-of "b")
+]
+
+Modify the mutable string-dict so that each key is now
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: No -- this returns a `List<b>`, and the description sounds like it returns a MSD.

File: src/trove/string-dict.scrbl [hunk 11]
```diff
@@ -423,4 +530,102 @@ check:
...
+associated with the result of applying
+the function @pyret{f} to its original value.
+
+@examples{
+fun one-of(ans, elts):
+  is-some(for find(elt from elts):
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: ditto

File: src/trove/string-dict.scrbl [hunk 12]
```diff
@@ -423,4 +530,102 @@ check:
...
+end
+}
+
+@msd-method["each-key-now"
+  #:contract (a-arrow (MSD-of "a") (a-arrow S No) No)
+  #:args (list (list "self" #f) (list "f" #f))
+  #:return No
+]
+
+Calls the function @pyret{f} on each key in the mutable string-dict,
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: ditto

Comment (ds26gte) [reply src/trove/string-dict.scrbl]: Thanks -- made your suggested changes. When having the description of `each-key` refer to `map-keys`, I tried using `@pyret-id{map-keys}` rather than `@pyret{map-keys}` to make it a hyperlink, but it doesn't seem to work. Is there a Pyret scribble quickref that will show how to ensure that `@pyret-id{...}` does link.

Comment (blerner) [reply src/trove/string-dict.scrbl]: Use `@pyret-method["StringDict"]{each-key}` to link to the `each-key` method of the `StringDict` data type.

File: src/trove/string-dict.scrbl [hunk 13]
```diff
@@ -281,18 +285,12 @@ end
...
-fun one-of(ans, elts):
-  is-some(for find(elt from elts):
-    ans == elt
-  end)
-end
-
 check:
   sd2 = [string-dict: "a", 10, "b", 6]
-  sd2.map-keys(lam(x): string-append(x, x) end) is%(one-of)
+  sd2.map-keys(lam(x): string-append(x, x) end) is%(lam(lft, rt): rt.member(lft) end)
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: This is complicated-looking.  Why not use `.sort()` like you did on line 273 above for `keys-list`?

File: src/trove/string-dict.scrbl [hunk 14]
```diff
@@ -576,20 +580,13 @@ end
...
-fun one-of(ans, elts):
-  is-some(for find(elt from elts):
-    ans == elt
-  end)
-end
-
 check:
   msd2 = [mutable-string-dict: "a", 10, "b", 6]
-  msd2.map-keys-now(lam(x): string-append(x, x) end) is%(one-of)
+  msd2.map-keys-now(lam(x): string-append(x, x) end) is%(lam(lft, rt): rt.member(lft) end)
```
Comment (blerner) [inline src/trove/string-dict.scrbl]: Ditto

**Comment by schanzer** [comment]
@ds26gte this PR seems awfully close to being mergeable. Any chance you can put an hour towards getting it across the finish line?
